homeassistant/components/binary_sensor/tod.py

```diff
@@ -0,0 +1,221 @@
+    STATE_OFF, STATE_ON, SUN_EVENT_SUNRISE, SUN_EVENT_SUNSET)
+from homeassistant.core import callback
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entity import Entity, async_generate_entity_id
+from homeassistant.helpers.event import async_track_point_in_time
+from homeassistant.helpers.sun import (
+    get_astral_event_date, get_astral_event_next)
+from homeassistant.util import dt as dt_util
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_AFTER_OFFSET = 'after_offset'
+CONF_BEFORE_OFFSET = 'before_offset'
+
+ATTR_NEXT_UPDATE = 'next_update'
+
+
+SENSOR_SCHEMA = vol.Schema({
+    vol.Required(CONF_AFTER): vol.Any(cv.time, vol.All(
+        vol.Lower, vol.Any(cv.sun_event))),
+    vol.Optional(CONF_AFTER_OFFSET, default=timedelta(0)): cv.time_period,
+    vol.Required(CONF_BEFORE): vol.Any(cv.time, vol.All(
+        vol.Lower, vol.Any(cv.sun_event))),
+    vol.Optional(CONF_BEFORE_OFFSET, default=timedelta(0)): cv.time_period,
+    vol.Optional(ATTR_FRIENDLY_NAME): cv.string,
+    vol.Optional(ATTR_ENTITY_ID): cv.entity_ids,
+})
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_SENSORS): vol.Schema({cv.slug: SENSOR_SCHEMA}),
+})
+
+
+async def async_setup_platform(hass, config, async_add_entities,
+                               discovery_info=None):
+    """Set up the ToD sensors."""
+    if hass.config.time_zone is None:
+        _LOGGER.error("Timezone is not set in Home Assistant configuration")
+        return False
+
+    sensors = []
+    for device, device_config in config[CONF_SENSORS].items():
+        after = device_config[CONF_AFTER]
+        after_offset = device_config[CONF_AFTER_OFFSET]
+        before = device_config[CONF_BEFORE]
+        before_offset = device_config[CONF_BEFORE_OFFSET]
+        sensors.append(
+            TodSensor(
+                hass, device, after, after_offset, before, before_offset
+            )
+        )
+
+    if not sensors:
+        _LOGGER.error("No sensors added")
+        return
+
+    async_add_entities(sensors)
+    return
+
+
+def is_sun_event(event):
+    """Return true if event is sun event not time."""
+    return event in (SUN_EVENT_SUNRISE, SUN_EVENT_SUNSET)
+
+
+class TodSensor(Entity):
+    """Time of the Day Sensor."""
+
+    def __init__(self, hass, sensor_name, after, after_offset,
+                 before, before_offset):
+        """Init the ToD Sensor..."""
+        self._name = sensor_name
+        self._state = None
+        self.hass = hass
+
+        self._time_before = None
+        self._time_after = None
+        self._after_offset = after_offset
+        self._before_offset = before_offset
+        self._before = before
+        self._after = after
+
+        self._sunrising = None
+        self._sunsetting = None
+        self._next_sunrising = None
+        self._next_sunsetting = None
+
+        self.entity_id = async_generate_entity_id(
+            ENTITY_ID_FORMAT, 'tod_{}'.format(sensor_name), hass=hass)
```

Comment:
Let's use the built-in entity id generator. Prefixing the entity is by `tod_` is unexpected.

Reply:
You mean just pass the sensor_name to generate_entity_id without prepending with tod_. Makes sense to avoid confusion as the name is provided in config.

Reply:
Well now that you're using the default entity ID format, there's no need for adding this explicitly. Home Assistant will automatically generate the entity ID for you.

Please remove `async_generate_entity_id` here. It will automatically be set.

Reply:
OK, Understand removed async_generate_entity_id and replaced with: 
`self.entity_id = ENTITY_ID_FORMAT.format(name)`

Reply:
No, I mean removing this entire line. HA will automatically set the entity ID, no need to assign it yourself

Reply:
Why? I have the sensor ids in the config. I don't want entity_id to be set from friendly name? Is that wrong?

```
binary_sensor:
  - platform: tod
    sensors:
      early_morning:  # I want this to be entity Id 
        after: sunrise
        after_offset: '-02:00'
        before: '07:00'
        friendly_name: "Wszesny poranek"
      morning:
        after: sunrise
        before: '12:00'
```

Reply:
Yes it is correct for template binary sensors. However, we decided some time ago to move to a new configuration format for new platforms like this. Please see the other comments I've left here, for example this one: https://github.com/home-assistant/home-assistant/pull/20068#issuecomment-453919773

Reply:
Is that the correct comment it?

Reply:
@OttoWinter Can you have a look, please:

```
1)

binary_sensor:
  - platform: tod
    sensors:
      early_morning:
        after: sunrise
        after_offset: '-02:00'
        before: '07:00'
        friendly_name: "Wszesny poranek"   # used local language on purpose
      morning:
        after: sunrise
        before: '12:00'
2)

binary_sensor:
  - platform: tod
    tods:
      - name: Wczesny poranek  # the entity_id will be automatically generated based on this
        after: sunrise
        after_offset: '-02:00'
        before: '07:00'
      - name: Morning
        after: sunrise
        before: '12:00'
```

So is my understanding correct you would recommend option 2 for that sensor?
IMHO option 1 is much easier to understand when you read it. It clearly indicates that there are multiple sensors created with this platforms. Also If English is not your native language and you still want to keep your entity_id in English, for sake of consistency with the rest of you config, and friendly_name in local language, option 2 seems be better than generating the entity id based on the name which may be in other language.
I don't have enough experience yet, and maybe I can't see issues that are obvious to you. 
Would appreciate your comment and explanation. I'm not tight to any option yet, and willing to get into final consensus here and eventually close this PR.

Reply:
@OttoWinter Hi Otto, could you please take a look at my comments?
There could be also a third option:

```
3) 
binary_sensor:
  - platform: tod
     name: Wczesny poranek
     after: sunrise
     after_offset: '-02:00'
     before: '07:00'
  
  - platform: tod 
     name: Morning
     after: sunrise
     before: '12:00'
```

I'm a bit confused and It's not clear to me what option is preferred.

Reply:
@OttoWinter Any reply on this pls?

Reply:
@kstaniek Yes, option 3. The template sensor way of doing it all in `sensors:` was always very confusing to users (from community forum issue count)

Reply:
Updated the config and tests accordingly. Option 3 has been implemented for config.